### PR TITLE
Feature: transparent checkout boleto

### PR DESCRIPTION
**Pull request by lflimeira**
**Descrição**

The PR adds the feature to buy with the payment method Boleto in the transparent checkout mode.

**Testes Realizados**

Acceptance tests.

### CODE REVIEW

File: app/code/community/PagarMe/Boleto/Block/Info/Boleto.php [hunk 1]
```diff
@@ -0,0 +1,12 @@
+<?php
+
+class PagarMe_Boleto_Block_Info_Boleto extends Mage_Payment_Block_Info
```
Comment (henriquekano) [inline app/code/community/PagarMe/Boleto/Block/Info/Boleto.php]: This does not seem like a important class o.0

File: app/code/community/PagarMe/Boleto/Helper/Data.php [hunk 2]
```diff
@@ -0,0 +1,39 @@
+<?php
+
+class PagarMe_Boleto_Helper_Data extends Mage_Core_Helper_Abstract
+{
+    public function buildCustomer()
+    {
+        $quote = $this->getQuote();
```
Comment (henriquekano) [inline app/code/community/PagarMe/Boleto/Helper/Data.php]: I'm not sure this `$this->getQuote()` would work...
Even if it works, the `buildCustomer` is used in a lot of models, can't we make it reusable for everyone of them?

File: app/code/community/PagarMe/Boleto/Model/Boleto.php [hunk 3]
```diff
@@ -0,0 +1,219 @@
+<?php
+use \PagarMe\Sdk\PagarMe as PagarMeSdk;
+
+class PagarMe_Boleto_Model_Boleto extends Mage_Payment_Model_Method_Abstract
+{
+    protected $_code = 'pagarme_boleto';
+    protected $_formBlockType = 'pagarme_boleto/form_boleto';
+    protected $_infoBlockType = 'pagarme_boleto/info_boleto';
```
Comment (henriquekano) [inline app/code/community/PagarMe/Boleto/Model/Boleto.php]: See the above

File: app/code/community/PagarMe/Boleto/Model/Boleto.php [hunk 4]
```diff
@@ -0,0 +1,219 @@
...
+    }
+    
+    /**
+     * @param array $data
+     *
+     * @return $this
+     */
+    public function assignData($data)
+    {
+
```
Comment (henriquekano) [inline app/code/community/PagarMe/Boleto/Model/Boleto.php]: line

File: app/code/community/PagarMe/Boleto/Model/Boleto.php [hunk 5]
```diff
@@ -0,0 +1,219 @@
...
+            'pagarme_boleto_url' => $transaction->getBoletoUrl(),
+        ];
+        
+        return array_merge(
+            $infoInstance->getAdditionalInformation(),
+            $data
+        );
+    }
+
+    public function capture(Varien_Object $payment, $amount)
```
Comment (henriquekano) [inline app/code/community/PagarMe/Boleto/Model/Boleto.php]: I think this method isn't used...

File: app/code/community/PagarMe/Boleto/etc/config.xml [hunk 6]
```diff
@@ -0,0 +1,68 @@
+<?xml version="1.0"?>
+<config>
+    <modules>
+        <PagarMe_Boleto>
+            <version>3.2.0</version>
```
Comment (henriquekano) [inline app/code/community/PagarMe/Boleto/etc/config.xml]: Why 3.2.0?

File: app/code/community/PagarMe/Core/etc/config.xml [hunk 7]
```diff
@@ -103,9 +103,17 @@
                 <model>pagarme_creditcard/creditcard</model>
                 <order_status>pending</order_status>
                 <allowspecific>0</allowspecific>
-                <payment_methods>credit_card,boleto</payment_methods>
+                <payment_methods>credit_card</payment_methods>
                 <payment_action>authorize</payment_action>
             </pagarme_creditcard>
+            <pagarme_boleto>
```
Comment (henriquekano) [inline app/code/community/PagarMe/Core/etc/config.xml]: shouldn't this be in the `PagarMe_Boleto`'s `config.xml`?

File: app/design/frontend/base/default/layout/pagarme/pagarme_modal.xml [hunk 8]
```diff
@@ -11,7 +11,7 @@
     </default>
     <checkout_onepage_success>
         <reference name="content">
-            <block type="pagarme_modal/success" name="pagarme_modal.success" template="pagarme/success.phtml"/>
+          <block type="pagarme_modal/success" name="pagarme_modal.success" template="pagarme/boleto/success.phtml"/>
```
Comment (henriquekano) [inline app/design/frontend/base/default/layout/pagarme/pagarme_modal.xml]: indentation

File: app/locale/pt_BR/PagarMe_Boleto.csv [hunk 9]
```diff
@@ -0,0 +1,29 @@
+"Pagar.me - Settings","Pagar.me - Configurações"
```
Comment (henriquekano) [inline app/locale/pt_BR/PagarMe_Boleto.csv]: why so many messages?

**Comment by lflimeira** [comment]
@henriquekano the icons.png is because in the checkout I show de boleto's image when the client chooses the payment method.

File: app/code/community/PagarMe/Core/etc/config.xml (hunk 7 above)
Comment (lflimeira) [reply app/code/community/PagarMe/Core/etc/config.xml]: It has to be in this config file, without it the Magento don't recognize and call the boleto's model class.

File: app/code/community/PagarMe/Boleto/etc/config.xml (hunk 6 above)
Comment (lflimeira) [reply app/code/community/PagarMe/Boleto/etc/config.xml]: Done.

**Comment by henriquekano** [comment]
lol

File: app/design/frontend/base/default/layout/pagarme/pagarme_modal.xml (hunk 8 above)
Comment (lflimeira) [reply app/design/frontend/base/default/layout/pagarme/pagarme_modal.xml]: Done.

File: app/locale/pt_BR/PagarMe_Boleto.csv (hunk 9 above)
Comment (lflimeira) [reply app/locale/pt_BR/PagarMe_Boleto.csv]: Done.

File: app/code/community/PagarMe/Boleto/Model/Boleto.php (hunk 5 above)
Comment (lflimeira) [reply app/code/community/PagarMe/Boleto/Model/Boleto.php]: done.

File: app/code/community/PagarMe/Boleto/Model/Boleto.php (hunk 4 above)
Comment (lflimeira) [reply app/code/community/PagarMe/Boleto/Model/Boleto.php]: done.

File: app/code/community/PagarMe/Boleto/Model/Boleto.php (hunk 3 above)
Comment (lflimeira) [reply app/code/community/PagarMe/Boleto/Model/Boleto.php]: Done.

File: app/code/community/PagarMe/Boleto/Helper/Data.php (hunk 2 above)
Comment (lflimeira) [reply app/code/community/PagarMe/Boleto/Helper/Data.php]: What do you propose?

Comment (henriquekano) [reply app/code/community/PagarMe/Boleto/Helper/Data.php]: I'd put it in the core helper, make a simple class or make a trait out of it

File: app/code/community/PagarMe/Core/etc/config.xml (hunk 7 above)
Comment (henriquekano) [reply app/code/community/PagarMe/Core/etc/config.xml]: And isn't that what was supposed to happen?

File: app/code/community/PagarMe/Boleto/Block/Success.php [hunk 10]
```diff
@@ -0,0 +1,58 @@
...
+    }
+
+    /**
+     * @return bool
+     */
+    public function isBoletoPayment()
+    {
+        $order = $this->getOrder();
+        $additionalInfo = $order->getPayment()->getAdditionalInformation();
+        $paymentMethod = (array_key_exists('pagarme_payment_method', $additionalInfo)) ? $additionalInfo['pagarme_payment_method'] : null;
```
Comment (leonampd) [inline app/code/community/PagarMe/Boleto/Block/Success.php]: Why we need this check?

if the `pagarme_payment_method` doesn't exists `$paymentMethod` will receive `null`, right? So, why not:

```
$paymentMethod = null;
if(array_key_exists('pagarme_payment_method', $additionalInformation) {
    $paymentMethod = $additionalInformation['pagarme_payment_method'];
}

File: app/code/community/PagarMe/Modal/Block/Success.php [hunk 11]
```diff
@@ -38,8 +38,10 @@ public function isBoletoPayment()
         $order = $this->getOrder();
 
         $additionalInfo = $order->getPayment()->getAdditionalInformation();
-
-        if ($additionalInfo['pagarme_payment_method'] === PagarMe_Modal_Model_Modal::PAGARME_MODAL_BOLETO) {
+        
+        $paymentMethod = (array_key_exists('pagarme_payment_method', $additionalInfo)) ? $additionalInfo['pagarme_payment_method'] : null;
```
Comment (leonampd) [inline app/code/community/PagarMe/Modal/Block/Success.php]: This is the second time that I see this piece of code. Can we extract this within a method?

Comment (leonampd) [reply app/code/community/PagarMe/Modal/Block/Success.php]: Also, ternary operations isn't readable. I prefer to use this only in value attribution cases. In this piece of code there's a verification too and, IMHO it make the code difficult to understand

File: app/code/community/PagarMe/Boleto/Block/Success.php (hunk 10 above)
Comment (henriquekano) [reply app/code/community/PagarMe/Boleto/Block/Success.php]: The point of this check is to not access the `pagarme_payment_method` key directly. When the key doesn't exists, a log will be printed (either in the system.log or exception.log, don't remember). By checking it first, we keep the logs a little cleaner.

Comment (leonampd) [reply app/code/community/PagarMe/Boleto/Block/Success.php]: I think it's difficult to read/understand. How about my example? I think it's more cleaner.

Comment (henriquekano) [reply app/code/community/PagarMe/Boleto/Block/Success.php]: LGTM

File: app/code/community/PagarMe/Boleto/Helper/Data.php (hunk 2 above)
Comment (lflimeira) [reply app/code/community/PagarMe/Boleto/Helper/Data.php]: Done.

File: app/code/community/PagarMe/Modal/Block/Success.php [hunk 12]
```diff
@@ -38,8 +38,10 @@ public function isBoletoPayment()
         $order = $this->getOrder();
 
         $additionalInfo = $order->getPayment()->getAdditionalInformation();
-
-        if ($additionalInfo['pagarme_payment_method'] === PagarMe_Modal_Model_Modal::PAGARME_MODAL_BOLETO) {
+        
+        $paymentMethod = (array_key_exists('pagarme_payment_method', $additionalInfo)) ? $additionalInfo['pagarme_payment_method'] : null; 
+        $isBoleto = $paymentMethod === PagarMe_Modal_Model_Modal::PAGARME_MODAL_BOLETO;
+        if ($isBoleto) {
```
Comment (leonampd) [inline app/code/community/PagarMe/Modal/Block/Success.php]: Why attribute a conditional to a var? You can put that in the if statment directly. 😉

Comment (lflimeira) [reply app/code/community/PagarMe/Modal/Block/Success.php]: Done

**Comment by henriquekano** [comment]
Is there any reason to keep both the modal's and the boleto's `success.phtml` templates?

File: app/code/community/PagarMe/Boleto/Block/Success.php [hunk 13]
```diff
@@ -0,0 +1,60 @@
...
+            );
+        }
+
+        return $this->order;
+    }
+
+    /**
+     * @codeCoverageIgnore
+     */
+    public function setOrder(Mage_Sales_Model_Order $order)
```
Comment (henriquekano) [inline app/code/community/PagarMe/Boleto/Block/Success.php]: When is this method used?

**Comment by lflimeira** [comment]
Yes @henriquekano, the payment method and that logic that we did to print the boleto button.

File: app/code/community/PagarMe/Boleto/etc/config.xml [hunk 14]
```diff
@@ -0,0 +1,68 @@
+<?xml version="1.0"?>
+<config>
+    <modules>
+        <PagarMe_Boleto>
+            <version>1.0.0</version>
```
Comment (henriquekano) [inline app/code/community/PagarMe/Boleto/etc/config.xml]: This version should match all the other modules'

File: app/code/community/PagarMe/Boleto/etc/config.xml [hunk 15]
```diff
@@ -0,0 +1,68 @@
...
+                <pagarme_boleto>
+                    <file>pagarme/pagarme_boleto.xml</file>
+                </pagarme_boleto>
+            </updates>
+        </layout>
+    </frontend>
+    <frontend>
+      <translate>
+        <modules>
+          <consultoras>
```
Comment (henriquekano) [inline app/code/community/PagarMe/Boleto/etc/config.xml]: wth is this `consultoras`?

File: app/code/community/PagarMe/Boleto/etc/config.xml [hunk 16]
```diff
@@ -0,0 +1,68 @@
...
+              <default>PagarMe_Boleto.csv</default>
+            </files>
+          </consultoras>
+        </modules>
+      </translate>
+    </frontend>
+    <admin>
+      <translate>
+        <modules>
+          <consultoras>
```
Comment (henriquekano) [inline app/code/community/PagarMe/Boleto/etc/config.xml]: Same as above

File: app/code/community/PagarMe/Boleto/Block/Success.php (hunk 13 above)
Comment (lflimeira) [reply app/code/community/PagarMe/Boleto/Block/Success.php]: Never :rofl: / Done

File: app/code/community/PagarMe/Boleto/etc/config.xml (hunk 14 above)
Comment (lflimeira) [reply app/code/community/PagarMe/Boleto/etc/config.xml]: Yep, we defined it today. Done.

Comment (lflimeira) [reply app/code/community/PagarMe/Boleto/etc/config.xml]: done.

File: app/code/community/PagarMe/Boleto/etc/config.xml (hunk 15 above)
Comment (lflimeira) [reply app/code/community/PagarMe/Boleto/etc/config.xml]: done.

File: app/code/community/PagarMe/Boleto/etc/config.xml (hunk 16 above)
Comment (lflimeira) [reply app/code/community/PagarMe/Boleto/etc/config.xml]: done.

File: app/code/community/PagarMe/Boleto/etc/config.xml [hunk 17]
```diff
@@ -0,0 +1,68 @@
...
+                <pagarme_boleto>
+                    <file>pagarme/pagarme_boleto.xml</file>
+                </pagarme_boleto>
+            </updates>
+        </layout>
+    </frontend>
+    <frontend>
+      <translate>
+        <modules>
+          <boleto>
```
Comment (henriquekano) [inline app/code/community/PagarMe/Boleto/etc/config.xml]: Should this be just the `module` name or `vendor_module`?

Comment (lflimeira) [reply app/code/community/PagarMe/Boleto/etc/config.xml]: AFAIK is just the translate module's name. Would be better name as pagarme_boleto?

Comment (henriquekano) [reply app/code/community/PagarMe/Boleto/etc/config.xml]: ¯\_(ツ)_/¯

Comment (henriquekano) [reply app/code/community/PagarMe/Boleto/etc/config.xml]: ¯\\_(ツ)_/¯

Comment (lflimeira) [reply app/code/community/PagarMe/Boleto/etc/config.xml]: Done.

Comment (henriquekano) [reply app/code/community/PagarMe/Boleto/etc/config.xml]: BTW, the indentation looks wrong

**Comment by henriquekano** [comment]
Now the core's config.xml and the boleto's config.xml indentation are divergent. The correct one is, AFAIK, 4 spaces.

File: tests/acceptance/BoletoContext.php [hunk 18]
```diff
@@ -0,0 +1,190 @@
...
+                Mage::helper('pagarme_boleto')
+                ->__('Place Order')
+            );
+    }
+    /**
+     * @Then the purchase must be paid with success
+     */
+    public function thePurchaseMustBePaidWithSuccess()
+    {
+        $this->session->wait(10000);
```
[the remark this answers is not in the document]
Comment (lflimeira) [reply tests/acceptance/BoletoContext.php]: Done.

**Pull request merged**